File: src/main/java/phoenix/init/events/PhoenixCommonEvents.java
package phoenix.init.events;

import net.minecraft.block.FlowingFluidBlock;
import net.minecraft.entity.EntityClassification;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.biome.Biomes;
import net.minecraft.world.gen.GenerationStage;
import net.minecraft.world.gen.feature.Feature;
import net.minecraft.world.gen.feature.IFeatureConfig;
import net.minecraft.world.gen.feature.OreFeatureConfig;
import net.minecraft.world.gen.feature.structure.Structure;
import net.minecraft.world.gen.placement.CountRangeConfig;
import net.minecraft.world.gen.placement.IPlacementConfig;
import net.minecraft.world.gen.placement.Placement;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.IForgeRegistry;
import phoenix.Phoenix;
import phoenix.init.*;
import phoenix.network.NetworkHandler;
import phoenix.utils.block.INonItem;

@Mod.EventBusSubscriber(modid = Phoenix.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
public class PhoenixCommonEvents
{
    @SubscribeEvent
    public static void onRegisterItems(final RegistryEvent.Register<Item> event)
    {
        final IForgeRegistry<Item> registry = event.getRegistry();
        PhoenixBlocks.getBLOCKS().getEntries().stream()
                .map(RegistryObject::get)
                .filter(block -> !(block instanceof INonItem))
                .filter(block -> !(block instanceof FlowingFluidBlock))
                .forEach(block ->
                {
                    final Item.Properties prop = new Item.Properties().group(Phoenix.Companion.getPHOENIX());
                    final BlockItem blockItem = new BlockItem(block, prop);
                    blockItem.setRegistryName(block.getRegistryName());
                    registry.register(blockItem);
                });
    }

    @SubscribeEvent
    public static void init(FMLCommonSetupEvent event)
    {
        PhoenixSounds.init();
        NetworkHandler.init();
        FMLJavaModLoadingContext.get().getModEventBus().register(PhoenixCommonEvents.class);
        PhoenixRecipes.register();

        PhoenixBiomes.getUNDER()    .get().addStructure(PhoenixFeatures.ERASED.get().withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG));
        PhoenixBiomes.getHEARTVOID().get().addStructure(PhoenixFeatures.ERASED.get().withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG));
        PhoenixBiomes.getUNDER()    .get().addFeature(GenerationStage.Decoration.SURFACE_STRUCTURES, PhoenixFeatures.ERASED.get().withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG).withPlacement(Placement.NOPE.configure(IPlacementConfig.NO_PLACEMENT_CONFIG)));
        PhoenixBiomes.getHEARTVOID().get().addFeature(GenerationStage.Decoration.SURFACE_STRUCTURES, PhoenixFeatures.ERASED.get().withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG).withPlacement(Placement.NOPE.configure(IPlacementConfig.NO_PLACEMENT_CONFIG)));

        PhoenixBiomes.getUNDER()    .get().addSpawn(EntityClassification.CREATURE, new Biome.SpawnListEntry(PhoenixEntities.getCAUDA().get(), 15, 1, 3));
        PhoenixBiomes.getHEARTVOID().get().addSpawn(EntityClassification.CREATURE, new Biome.SpawnListEntry(PhoenixEntities.getTALPA().get(), 15, 1, 4));

        addStructure(Biomes.END_HIGHLANDS, PhoenixFeatures.ERASED.get());
        addStructure(Biomes.END_HIGHLANDS, PhoenixFeatures.CORN.get());

        for (Biome biome : Registry.BIOME)
        {
            if(biome != Biomes.END_BARRENS && biome != Biomes.END_HIGHLANDS && biome != Biomes.END_MIDLANDS && biome != Biomes.THE_END &&
                    biome != Biomes.SMALL_END_ISLANDS && biome != PhoenixBiomes.getUNDER().get() && biome != PhoenixBiomes.getHEARTVOID().get())
            {
                addZirconiumOre(biome);
            }
        }

        //Minecraft.getInstance().particles.registerFactory(((ParticleType)PhoenixParticles.INSTANCE.getPHOENIX_BORN().get()), new PhoenixBornFactory());
    }

    public static void addStructure(Biome biome, Structure structure)
    {
        biome.addStructure(structure.withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG));
        biome.addFeature(GenerationStage.Decoration.SURFACE_STRUCTURES, structure.withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG)
                .withPlacement(Placement.NOPE.configure(IPlacementConfig.NO_PLACEMENT_CONFIG)));
    }

    public static void addZirconiumOre(Biome biome)
    {
        biome.addFeature(GenerationStage.Decoration.UNDERGROUND_ORES,
                Feature.ORE.withConfiguration(new OreFeatureConfig(OreFeatureConfig.FillerBlockType.NATURAL_STONE, PhoenixBlocks.getZIRCONIUM().get().getDefaultState(), 4))
                        .withPlacement(Placement.COUNT_RANGE.configure(new CountRangeConfig(20, 0, 0, 64))));
    }
}

File: src/main/kotlin/phoenix/init/events/PhoenixCommonEvents.kt
package phoenix.init.events

import net.minecraft.block.Block
import net.minecraft.block.FlowingFluidBlock
import net.minecraft.block.material.MaterialColor
import net.minecraft.entity.EntityClassification
import net.minecraft.item.*
import net.minecraft.item.crafting.Ingredient
import net.minecraft.potion.PotionUtils
import net.minecraft.potion.Potions.MUNDANE
import net.minecraft.util.ResourceLocation
import net.minecraft.util.registry.Registry
import net.minecraft.world.biome.Biome.SpawnListEntry
import net.minecraft.world.biome.Biomes.*
import net.minecraft.world.gen.GenerationStage
import net.minecraft.world.gen.feature.IFeatureConfig
import net.minecraftforge.common.brewing.BrewingRecipeRegistry
import net.minecraftforge.event.RegistryEvent.Register
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod.EventBusSubscriber
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.registries.ForgeRegistries
import phoenix.MOD_ID
import phoenix.Phoenix
import phoenix.Phoenix.Companion.ASH
import phoenix.api.block.ICustomGroup
import phoenix.api.block.INonItem
import phoenix.init.PhxBiomes.HEARTVOID
import phoenix.init.PhxBiomes.UNDER
import phoenix.init.PhxBlocks.blocks
import phoenix.init.PhxEntities
import phoenix.init.PhxEntities.cauda
import phoenix.init.PhxEntities.talpa
import phoenix.init.PhxFeatures
import phoenix.init.PhxItems
import phoenix.init.PhxPotions.LEVITATION
import phoenix.init.PhxPotions.LONG_LEVITATION
import phoenix.init.PhxRecipes
import phoenix.items.FixedSpawnEggItem
import phoenix.network.initPacketSystem
import phoenix.other.addStructure
import phoenix.other.addZirconiumOre
import phoenix.utils.ServerStageUppedEvent
import thedarkcolour.kotlinforforge.forge.ObjectHolderDelegate

@EventBusSubscriber(modid = MOD_ID, bus = EventBusSubscriber.Bus.MOD)
object PhoenixCommonEvents
{
    @SubscribeEvent
    fun onRegisterItems(event: Register<Item>)
    {
        blocks.getEntries().stream()
                .map(ObjectHolderDelegate<out Block>::get)
                .filter { block: Block -> block !is INonItem }
                .filter { block: Block -> block !is FlowingFluidBlock }
                .forEach { block: Block ->
                    val tab = if(block is ICustomGroup) block.tab else ASH
                    val prop = Item.Properties().group(tab)
                    val blockItem = BlockItem(block, prop)
                    blockItem.registryName = block.registryName
                    event.registry.register(blockItem)
                }
    }

    @SubscribeEvent
    fun init(event: FMLCommonSetupEvent)
    {
        initPacketSystem()
        PhxRecipes.register()
        UNDER.addSpawn(EntityClassification.CREATURE, SpawnListEntry(cauda, 15, 1, 1))
        HEARTVOID.addSpawn(EntityClassification.CREATURE, SpawnListEntry(talpa, 15, 1, 4))
        END_HIGHLANDS.addStructure(PhxFeatures.REMAINS)
        UNDER.addStructure(PhxFeatures.REMAINS)
        UNDER.addFeature(GenerationStage.Decoration.UNDERGROUND_DECORATION, PhxFeatures.WATER.withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG))
        HEARTVOID.addFeature(GenerationStage.Decoration.UNDERGROUND_DECORATION, PhxFeatures.WET_TREE.withConfiguration(IFeatureConfig.NO_FEATURE_CONFIG))

        for (biome in Registry.BIOME)
        {
            if (biome !== END_BARRENS && biome !== END_HIGHLANDS && biome !== END_MIDLANDS && biome !== THE_END &&
                    biome !== SMALL_END_ISLANDS && biome !== UNDER && biome !== HEARTVOID && biome !== THE_VOID && biome !== NETHER)
            {
                biome.addZirconiumOre()
            }
        }

        BrewingRecipeRegistry.addRecipe(
            Ingredient.fromStacks(PotionUtils.addPotionToItemStack(ItemStack(Items.POTION), MUNDANE)),
            Ingredient.fromItems(PhxItems.GOLDEN_SETA),
            PotionUtils.addPotionToItemStack(ItemStack(Items.POTION), LEVITATION)
        )

        BrewingRecipeRegistry.addRecipe(
            Ingredient.fromStacks(PotionUtils.addPotionToItemStack(ItemStack(Items.POTION), LEVITATION)),
            Ingredient.fromItems(Items.REDSTONE),
            PotionUtils.addPotionToItemStack(ItemStack(Items.POTION), LONG_LEVITATION)
        )
    }

    @SubscribeEvent
    fun registerItems(event : Register<Item>)
    {
        if (event.registry === ForgeRegistries.ITEMS)
        {
            val caudaEgg = FixedSpawnEggItem(PhxEntities::cauda, MaterialColor.BLACK.colorValue, MaterialColor.BLACK_TERRACOTTA.colorValue, Item.Properties().group(ItemGroup.MISC))
            caudaEgg.registryName = ResourceLocation(MOD_ID, "cauda_spawn_egg")
            val golemSpawnEntity = FixedSpawnEggItem(PhxEntities::ancientGolemEntity, MaterialColor.CYAN.colorValue, MaterialColor.BLACK.colorValue, Item.Properties().group(ItemGroup.MISC))
            golemSpawnEntity.registryName = ResourceLocation(MOD_ID, "ancient_golem_spawn_egg")
            ForgeRegistries.ITEMS.registerAll(caudaEgg, golemSpawnEntity)
        }
    }

    @SubscribeEvent
    fun onStageUppded(event : ServerStageUppedEvent)
    {
    }
}